modules/storage/storage.go

```diff
@@ -0,0 +1,287 @@
+
+	return s.ensureCache()
+}
+
+// CreateFilesystem with the given size in a storage pool.
+func (s *storageModule) CreateFilesystem(size uint64) (string, error) {
+	log.Info().Msgf("Creating new volume with size %d", size)
+
+	fs, err := s.createFs(size, uuid.New().String(), false)
+	if err != nil {
+		return "", err
+	}
+	return fs.Path(), nil
+}
+
+// ReleaseFilesystem at the given path, this will unmount and then delete
+// the filesystem. After this call, the caller must not perform any more actions
+// on this filesystem
+func (s *storageModule) ReleaseFilesystem(path string) error {
+	log.Info().Msgf("Deleting volume at %v", path)
+
+	for _, volume := range s.volumes {
+		filesystems, err := volume.Volumes()
+		if err != nil {
+			return err
+		}
+		for _, fs := range filesystems {
+			if fs.Path() == path {
+				log.Debug().Msgf("Removing filesystem %v in volume %v", fs.Name(), volume.Name())
+				return volume.RemoveVolume(fs.Name())
+			}
+		}
+	}
+
+	log.Warn().Msgf("Could not find filesystem %v", path)
+	return nil
+}
+
+// ensureCache creates a "cache" subvolume and mounts it in /var
+func (s *storageModule) ensureCache() error {
+	log.Info().Msgf("Setting up cache")
+
+	log.Debug().Msgf("Checking pools for existing cache")
+
+	var cacheFs filesystem.Volume
+
+	// check if we already have a cache
+	for _, volume := range s.volumes {
+		filesystems, err := volume.Volumes()
+		if err != nil {
+			return err
+		}
+		for _, fs := range filesystems {
+			if fs.Name() == cacheLabel {
+				log.Debug().Msgf("Found existing cache at %v", fs.Path())
+				cacheFs = fs
+				break
+			}
+		}
+		if cacheFs != nil {
+			break
+		}
+	}
+
+	if cacheFs == nil {
+		log.Debug().Msgf("No cache found, try to create new cache")
+
+		fs, err := s.createFs(cacheSize, cacheLabel, true)
+		if err != nil {
+			return err
+		}
+		cacheFs = fs
+	}
+
+	log.Debug().Msgf("Mounting cache partition in %s", cacheTarget)
+	return filesystem.BindMount(cacheFs, cacheTarget)
+}
+
+// createFs creates a filesystem with the given name and limits it to the given size
+// if the requested disk type does not have a storage pool available, an error is
+// returned
+func (s *storageModule) createFs(size uint64, name string, preferSSD bool) (filesystem.Volume, error) {
+	var fs filesystem.Volume
+	var err error
+
+	possiblePools := s.volumes
+	if preferSSD {
+		ssdPools := []filesystem.Pool{}
+		hddPools := []filesystem.Pool{}
+		for _, pool := range s.volumes {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+			defer cancel()
+
+			poolType, err := s.devices.PoolType(ctx, pool)
+			if err != nil {
+				return nil, err
+			}
+			if poolType == filesystem.SSDDevice {
+				ssdPools = append(ssdPools, pool)
+			} else {
+				hddPools = append(hddPools, pool)
+			}
+		}
+		possiblePools = append(ssdPools, hddPools...)
+	}
+
+	// for now take the first volume, if an ssd is prefered ssds will be sorted
+	// first
+	for _, volume := range possiblePools {
```

Comment:
I don't see where we verify that the selected volume still have enough free space.